archiver/archiveCrawlers.ts

```diff
@@ -0,0 +1,138 @@
+
+const mkdirp = async (filePath: string) => {
+  try {
+    // Ensure dirs exists with mkdir
+    mkdirSync(filePath, { recursive: true });
+  } catch (error) {
+    // noop. Throws error if dirs exists
+  }
+};
+
+const archiveJson = async (archivePath: string, showlistData: {}) => {
+  const archiveFilePath = path.join(archivePath, `showlist.json`);
+  await fs.writeFile(archiveFilePath, JSON.stringify(showlistData), 'utf-8');
+};
+
+const downloadFile = async (url: string, destinationPath: string) => {
+  // TODO: stream files directly to AWS?
+  const response = await fetch(url);
+  const arrayBuffer = await response.arrayBuffer();
+  const buffer = Buffer.from(arrayBuffer);
+
+  await fs.writeFile(destinationPath, buffer);
+};
+
+/** Contentful - Pre wappu 2023 */
+const ARCHIVE_SOURCE_URL = process.env.ARCHIVE_SOURCE_URL || '';
+
+const archiveContentful = async (showlistId: string, name: string) => {
+  const archivePath = getArchivePath(name);
+  const pictureFolder = path.join(archivePath, 'pictures');
+  mkdirp(archivePath);
+  mkdirp(pictureFolder);
+
+  const showlistResponse = await fetchContentfulContent<ShowlistQuery>(
+    ShowlistDocument,
+    { showlistId }
+  );
+  // NOTE: when data is stored locally
+  // const showlistResponse = raw as ShowlistQuery;
+  const shows = parseQueryResultToShowlist(showlistResponse);
+
+  const showsWithNewFilePaths = await reconnectShowsWithLocalFiles(
+    shows,
+    pictureFolder,
+    ARCHIVE_SOURCE_URL
+  );
+  const showlistData = showsToGroups(showsWithNewFilePaths);
+
+  await archiveJson(archivePath, showlistData);
+};
+
+const reconnectShowsWithLocalFiles = async (
+  shows: Show[],
+  archivePath: string,
+  archiveUrl: string
+): Promise<Show[]> => {
+  const showsWithLocalFiles = [];
+  for (const show of shows) {
+    const { name, pictureUrl } = show;
+    if (!pictureUrl) {
+      showsWithLocalFiles.push(show);
+      continue;
+    }
+
+
+    const extension = path.parse(pictureUrl).ext;
+    const newFilename = getImagePath('', name, extension);
+    const filePath = path.join(archivePath, newFilename);
+    const archiveFileUrl = path.join(archiveUrl, newFilename);
+    await downloadFile(pictureUrl, filePath);
+
+    showsWithLocalFiles.push({ ...show, pictureUrl: archiveFileUrl });
+  }
+  return showsWithLocalFiles;
+};
+
+/** Google Sheets - Post wappu 2023 */
+
+const archiveGogleSheet = async (
+  config: GoogleConfigSheets,
+  name: string,
+  showStartTime: string
+) => {
+  const archivePath = getArchivePath(name);
+  const pictureFolder = path.join(archivePath, 'pictures');
+  mkdirp(archivePath);
+  mkdirp(pictureFolder);
+
+  const showlistResponse = await getSheet(config);
+  const showlist = await parseSheetToShowList(showlistResponse, {
+    apiKey: process.env.GA_API_KEY,
+    fileUrlBase: path.join(ARCHIVE_SOURCE_URL, name, 'pictures'),
+    localFilePath: pictureFolder,
+    showStartTime,
+  });
+  await archiveJson(archivePath, showlist);
+};
```

Comment:
Mietin et voisko nää olla tuolla `scripts`-kansiossa mut ei kyl mitään väliä

Reply:
Ei voi olla, koska näitä ei voi ajaa ilman next-ympäristöä, koska graphql ja typescript vaatii omat buildstepit. Nyt ajellaan kömpelösti `/api/archive` rajapintaa pitkin, mutta lisään siihen apiin vielä jonkun envflagin, et ei voi edes yrittää pommittaa sitä rajapintaa